libs/secure/util.cpp

```diff
@@ -0,0 +1,73 @@
+/**
+ * Copyright Soramitsu Co., Ltd. 2017 All Rights Reserved.
+ * http://soramitsu.co.jp
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *        http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "secure/util.hpp"
+
+namespace util {
+  /**
+   * cast void ptr to volatile ptr of volatile T
+   */
+  template <typename T>
+  inline constexpr T volatile *volatile void2volatileptr(void *mem) {
+    return const_cast<volatile T *>(reinterpret_cast<T *>(mem));
+  }
+}
+
+void erase(void *dest, size_t size) {
+  char volatile *volatile p = util::void2volatileptr<char>(dest);
+
+  if (size == 0) {
+    return;
+  }
+
+  do {
+    *p = 0;
```

Comment:
Documentation says:
```
std::memset may be optimized away (under the as-if rules) if the object modified by this function is not accessed again for the rest of its lifetime
```

Which means under certain conditions it may not erase critical data.

Reply:
Also, volatile prevents compiler from any optimizations and critical data **guaranteed** to be erased.

Reply:
volatile + memset is never optimized